### PR TITLE
Overload `cholesky` method in order to prevent scalar indexing for `Diagonal`

### PR DESCRIPTION
**Pull request by simsurace**
Makes this work without scalar indexing:
```julia
using CUDA, LinearAlgebra
CUDA.allowscalar(false)

n = 1024
d = CUDA.rand(n)
D = Diagonal(d)
cholesky(D)
```
Performance can probably be improved.

### CODE REVIEW

**Comment by simsurace** [comment]
Unfortunately I have no experience with oneAPI.jl. Will look into the failing tests later, if people think that this is a useful PR.

**Comment by maleadt** [comment]
Thanks for the PR!

> Unfortunately I have no experience with oneAPI.jl. Will look into the failing tests later, if people think that this is a useful PR.

The failure is a generic GPU array failure where Base (specifically BLAS) functionality is called where a specialization is needed:

```
  ArgumentError: cannot take the host address of a oneArray{Float32, 2}
  Stacktrace:
    [1] unsafe_convert(#unused#::Type{Ptr{Float32}}, x::oneArray{Float32, 2})
      @ oneAPI ~/.cache/julia-buildkite-plugin/depots/c9f52312-b528-44e4-9501-6d408762012b/packages/oneAPI/KbEvp/src/array.jl:176
    [2] potrf!(uplo::Char, A::oneArray{Float32, 2})
      @ LinearAlgebra.LAPACK /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/lapack.jl:3088
    [3] _chol!
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/cholesky.jl:172 [inlined]
    [4] cholesky!(A::Hermitian{Float32, oneArray{Float32, 2}}, ::Val{false}; check::Bool)
      @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/cholesky.jl:252
    [5] cholesky!(A::oneArray{Float32, 2}, ::Val{false}; check::Bool)
      @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/cholesky.jl:285
    [6] #cholesky#134
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/cholesky.jl:378 [inlined]
    [7] cholesky (repeats 2 times)
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/cholesky.jl:378 [inlined]
    [8] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci3/julialang/gpuarrays-dot-jl/test/testsuite/linalg.jl:107 [inlined]
```

No GPUArrays in that stack trace, so maybe this PR is missing an overload (or relying on CUDA.jl-specific functionality, which isn't allowed for generic GPUArrays.jl implementations)?

**Comment by simsurace** [comment]
Ah, looks `cholesky` is generally broken for `oneArray`s.

**Comment by simsurace** [comment]
I think the reason for the failure is that the generic [`LinearAlgebra.LAPACK.potrf!` method](https://github.com/JuliaLang/julia/blob/35f0c911f443ae0d3933e1bce4321afd0f345aed/stdlib/LinearAlgebra/src/lapack.jl#L3078) that calls into BLAS does not work for `oneArray`s, a specific overload analogous to [the one in `CUDA/lib/cusolver/dense.jl`](https://github.com/JuliaGPU/CUDA.jl/blob/fff60874464f89ec1e82a25b90a9509166082c3e/lib/cusolver/dense.jl#L21) is missing. Shall I open an issue in oneAPI.jl? I presume the overload should be in there, not here.

**Comment by maleadt** [comment]
Those kind of overloads (replacing `LinearAlgebra.BLAS` and `LinearAlgebra.LAPACK` methods with GPU-specific ones) is something I'm trying to get rid of, actually, because CUBLAS really isn't a drop-in replacement for BLAS so it's a bad place for putting overrides.

That said, unless GPUArrays.jl can provide a generic `potrf` this is a bad test, so maybe it's better to move this implementation to CUDA.jl where the required functionality is available.

**Comment by simsurace** [comment]
So just to clarify: the generic `cholesky` overload introduced in this PR works on diagonal `oneArray`s and `CuArrays`. The problematic method was the one being called on dense `oneArray`s in the previous tests. I fixed the tests to not rely on these methods anymore. With this modification, everything passes.

File: test/testsuite/linalg.jl [hunk 1]
```diff
@@ -116,7 +104,7 @@
             n = 128
             d = AT(rand(Float32, n))
             D = Diagonal(d)
-            F = AT(collect(D))
+            F = collect(D)
```
Comment (simsurace) [inline test/testsuite/linalg.jl:107]: Prevent from calling dense `cholesky` method, which is not available for `oneArrays`

**Pull request merged**